src/sentry/tasks/process_buffer.py

```diff
@@ -17,19 +17,25 @@
 
 
 @instrumented_task(name='sentry.tasks.process_buffer.process_pending')
-def process_pending():
+def process_pending(shard=None):
     """
     Process pending buffers.
     """
     from sentry import buffer
     from sentry.app import locks
 
-    lock = locks.get('buffer:process_pending', duration=60)
+    if shard is None:
+        lock_key = 'buffer:process_pending'
+    else:
+        lock_key = 'buffer:process_pending:%d' % shard
+
+    lock = locks.get(lock_key, duration=60)
```

Comment:
Isn't this being locked twice now with different lock keys? Once here, and once in `process_pending`?

Reply:
Where does it lock inside `process_pending`? `process_pending` fans out to new Celery tasks. So this just created N locks that don't contend with each other. One per shard running in parallel.

Reply:
This entire block: https://github.com/getsentry/sentry/blob/24b2b091e6e6c893d02f6c7cf7193357e160aefc/src/sentry/buffer/redis.py#L142-L177

They don't contend with each other, but they are duplicated.

Reply:
Well, to be fair, this logic existed, and I didn't care to begin thinking about why it does it this way, but behavior shouldn't change from what it was before.

Reply:
Yep, didn't mean to say it was introduced here, I just found it while reading.

Reply:
imo the lock makes sense in its new form (and definitely in its old)

Reply:
I don't understand that comment?

Reply:
nvm